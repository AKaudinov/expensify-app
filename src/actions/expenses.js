import uuid from "uuid";
import fireBaseDb from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense: expense
    // expense: {
    //     // id: uuid(), //automatically set the id via the uuid library
    //     description: description,
    //     note: note,
    //     amount: amount,
    //     createdAt: createdAt
    // }
});


export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => { //redux-thunk allows to return a function that accepts a dispatch parameter, which allows to dispatch other actions
      const uid = getState().auth.uid;

      const {
          description = '',
          note = '',
          amount = 0,
          createdAt = 0
      } = expenseData; //destructure everything from the expenseData object

      const expense = {description, note, amount, createdAt};

      return fireBaseDb.ref(`users/${uid}/expenses`).push(expense).then(insertedExpenseRef => { //return the promise chain, so we can continue chaining if we want to
          dispatch(addExpense({
              id: insertedExpenseRef.key,
              ...expense
          }))
      }).catch(err => console.log(err));
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => { //default to empty object if the passed in object is undefined
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    }
};

//START_REMOVE_EXPENSE
export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
      const userId = getState().auth.uid;
      return fireBaseDb.ref(`users/${userId}/expenses/${id}`).remove()
          .then(() => {
              dispatch(removeExpense({id}))
          })
          .catch(err => {
              console.log(err);
          })
  }
};


//EDIT_EXPENSE
export const editExpense =(id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates
    }
};

//START_EDIT_EXPENSE
export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
      const userId = getState().auth.uid;
      return fireBaseDb.ref(`users/${userId}/expenses/${id}`).update({
          ...updates
      }).then(() => {
          dispatch(editExpense(id, updates))
      }).catch(err => {
          console.log(err);
      })
  }
};

//SET_EXPENSES //GET EXPENSES
export const getExpenses = (expenses) => {
    return {
        type: 'GET_EXPENSES',
        expenses: expenses
    }
};

export const startGetExpenses = () => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        const expenses = [];
        return fireBaseDb.ref(`users/${userId}/expenses`).once('value')
            //return the promise chain, so we can continue chaining in app.js, because it uses a then on `startGetExpenses` action
            .then(snapshot => {
                snapshot.forEach(fireBaseExpense => {
                    expenses.push({
                        id: fireBaseExpense.key,
                        ...fireBaseExpense.val()
                    })
                });
                dispatch(getExpenses(expenses))
            })
            .catch(err => console.log(err));
    };
};
