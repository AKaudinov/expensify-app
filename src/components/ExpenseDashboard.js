import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboard = (props) => (
    <div>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
);


export default ExpenseDashboard;
