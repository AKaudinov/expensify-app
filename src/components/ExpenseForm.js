import React from 'react';
import {connect} from 'react-redux'
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component{
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '', //convert the number to a number with cents by dividing by 100
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error:''
        };
    }

    onDescriptionChange = (e) =>{
      const description = e.target.value;
      this.setState(state => {
          return {
              description: description
          }
      });
    };

    onNoteChange = (e) => {
      const note = e.target.value;
      this.setState(prevState => {
         return {
             note: note //don't use e.target.value right in the callback without using e.persist(),
             //since this callback doesn't run right away, and by that time, e.target.value is lost
         }
      });
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){ //only see numbers that might have a decimal point, and a max of 2 numbers afterwards
            this.setState(prevState => ({
                amount: amount
            }));
        }
    };

    onDateChange = (createdAtMomentDate) => {
        if(createdAtMomentDate) { //if a data is actually passed in, set the state
            this.setState(prevState => {
                return {
                    createdAt: createdAtMomentDate
                }
            })
        }
    };

    onFocusChange = ({focused}) => { //destructure the first argument(focused) that gets passed in as part of an object
        this.setState(prevState => ({
            calendarFocused: focused
        }))
    };

    onSubmitForm = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            //set error state equal to 'Please provide description and amount'
            this.setState(prevState => {
                return{
                    error: 'Please provide description and amount'
                }
            })
        }else{
            //clear the error
            this.setState(prevState => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //format the number correctly
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render(){
        return (
            <div>
                {this.state.error && <h4>{this.state.error}</h4>}
              <form onSubmit={this.onSubmitForm}>
                  <input
                  type="text"
                  placeholder="Description"
                  autoFocus
                  value={this.state.description}
                  onChange={this.onDescriptionChange}/>

                  <input
                  type="text"
                  placeholder="Amount"
                  value={this.state.amount}
                  onChange={this.onAmountChange}/>

                  <SingleDatePicker
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  numberOfMonths={1}
                  isOutsideRange={day => false}/>

                  <textarea
                      placeholder="Add a note for your expense (optional)"
                      value={this.state.note}
                      onChange={this.onNoteChange}>
                  </textarea>

                  <button>Save expense</button>

              </form>
            </div>
        )
    }
}
