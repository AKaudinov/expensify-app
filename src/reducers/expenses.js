const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense); //concat creates a copy of an array or extends a previous array with a new one and returns a new array
            return [...state, action.expense]; //spread the items from previous array, to the new array,
        // and add a new object to the array, and return it. Old array doesn't get mutated
        case 'REMOVE_EXPENSE':
            // return state.filter(expense => expense.id !== action.id);
            return state.filter(({ id }) => id !== action.id); //filter the array returns a new array based on the filter function
        //in this case, return all the expenses that do not equal the provided id.
        //original state is not mutated
        case 'EDIT_EXPENSE':
            return state.map(expense => {
                if(expense.id === action.id){
                    return {
                        ...expense, //assign all the previous values to the new object
                        ...action.updates //update all the previously assigned values to the new object
                    };
                }else{
                    return expense;
                }
            }); //go through each item and modify the item that matches the provided id
        case 'GET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};
