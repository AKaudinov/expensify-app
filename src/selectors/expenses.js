import moment from "moment";

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter(expense => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
                                            //see if the start date filter is the same or starts before the expense has been created
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
                                            //see if the end date filter is the same or starts after the expense has been created
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        //startDateMatch and endDateMatch have to be in range of the created expense
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        //-1: comes first
        //1: b comes first
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1; //if a is less than b, b comes first, otherwise a comes first
        }else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1; //if a is less than b, b comes first, otherwise a comes first
        }
    });
};

export default getVisibleExpenses;
