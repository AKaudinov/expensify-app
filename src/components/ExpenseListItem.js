//export a stateless functional component
//description, amount, createdAt
import React from 'react';
import {Link} from "react-router-dom";
import {connect} from 'react-redux';
import * as ExpenseActions from '../actions/expenses';
import moment from 'moment';
import numeral from 'numeral';

//named export so it can be used for testing without being connected to redux store
export const ExpenseListItem = ({expense, dispatch}) => { //destructure expense from props, it comes in as props.expense
    return (
          <Link className="list-item" to={`edit/${expense.id}`}>
              <div>
                  <h3 className="list-item__title">{expense.description}</h3>
                  <span className="list-item__sub-title">created at: {moment(expense.createdAt).format('MM/DD/YYYY')}</span>
              </div>
              <h3 className="list-item__data">{numeral(expense.amount / 100).format('$0,0.00')}</h3>
              {/*convert expense.amount / 100 from cents to actual decimals*/}
          </Link>
    );
};

const connectedListItem = connect()(ExpenseListItem);
export default connectedListItem;
