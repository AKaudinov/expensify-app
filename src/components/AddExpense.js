import React from 'react';
import ExpenseForm from "./ExpenseForm";
import {connect} from 'react-redux';
import * as expenseActions from '../actions/expenses';

export class AddExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}/>
            </div>
        )
    }
}

// const AddExpense = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//         onSubmit={(expenseObj) => {
//             //props.dispatch(expenseActions.addExpense(expenseObj));
//             props.addExpense(expenseObj);
//             props.history.push('/'); //redirect back to the dashboard page, which in this case is home page
//         }}/>
//     </div>
// );

//similar to state to props
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(expenseActions.addExpense(expense))
    }
};

export default connect(undefined, mapDispatchToProps)(AddExpense);
