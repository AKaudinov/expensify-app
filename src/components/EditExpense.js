import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import * as expenseActions from '../actions/expenses';

export class EditExpense extends React.Component{
    onSubmit = (expense) => {
        this.props.editExpense(this.props.match.params.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
      this.props.removeExpense(this.props.expense);
      this.props.history.push('/');
    };

    render(){
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}/>
                <button onClick={this.onRemove}>Remove Expense</button>
            </div>
        );
    }
}

// const EditExpense = (props) => {
//     return (
//     <div>
//         <ExpenseForm
//             expense={props.expense}
//         onSubmit={expenseObj => {
//             props.dispatch(expenseActions.editExpense(props.match.params.id, expenseObj));
//             props.history.push('/');
//         }}/>
//         <button onClick={() => {
//             props.dispatch(expenseActions.removeExpense(props.expense));
//             props.history.push('/');
//         }}>Remove Expense</button>
//     </div>
//     );
// };

const mapStateToProps = (state, props) => { //find the expense with the given id from the passed in url
    return {
      expense: state.expenses.find(expenseItem => expenseItem.id === props.match.params.id) //search through the array and find the single item)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
      editExpense: (id, expense) => dispatch(expenseActions.editExpense(id, expense)),
        removeExpense: (expense) => dispatch(expenseActions.removeExpense(expense))
    }
};

const connectedEditExpense = connect(mapStateToProps, mapDispatchToProps)(EditExpense);
export default connectedEditExpense;
