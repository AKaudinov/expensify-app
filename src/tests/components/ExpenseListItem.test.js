import React from 'react';
import { shallow } from 'enzyme';
import {ExpenseListItem} from "../../components/ExpenseListItem"; //import the non-connected component version
import expenses from "../fixtures/expenses";

test('Should render ExpenseListItem', () => {
   const wrapper = shallow(<ExpenseListItem expense={expenses[0]}/>);
   expect(wrapper).toMatchSnapshot();
});
