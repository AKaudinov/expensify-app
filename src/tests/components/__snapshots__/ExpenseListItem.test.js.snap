// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem 1`] = `
<Link
  className="list-item"
  replace={false}
  to="edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      test1
    </h3>
    <span
      className="list-item__sub-title"
    >
      created at: 
      12/31/1969
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $1.95
  </h3>
</Link>
`;
